lighthouse-core/config/config.js

```diff
@@ -327,46 +330,31 @@ class Config {
       configJSON = Config.extendConfigJSON(deepCloneConfigJson(defaultConfig), configJSON);
     }
 
-    // Augment config with necessary defaults
-    configJSON = Config.augmentWithDefaults(configJSON);
-
-    // Expand audit/gatherer short-hand representations and merge in defaults
-    configJSON.audits = Config.expandAuditShorthandAndMergeOptions(configJSON.audits);
-    configJSON.passes = Config.expandGathererShorthandAndMergeOptions(configJSON.passes);
-
-    // Override any applicable settings with CLI flags
-    configJSON.settings = merge(configJSON.settings || {}, cleanFlagsForSettings(flags), true);
-
-    // Generate a limited config if specified
-    if (Array.isArray(configJSON.settings.onlyCategories) ||
-        Array.isArray(configJSON.settings.onlyAudits) ||
-        Array.isArray(configJSON.settings.skipAudits)) {
-      const categoryIds = configJSON.settings.onlyCategories;
-      const auditIds = configJSON.settings.onlyAudits;
-      const skipAuditIds = configJSON.settings.skipAudits;
-      configJSON = Config.generateNewFilteredConfig(configJSON, categoryIds, auditIds,
-          skipAuditIds);
-    }
+    // The directory of the config path, if one was provided.
+    const configDir = configPath ? path.dirname(configPath) : undefined;
 
-    Config.adjustDefaultPassForThrottling(configJSON);
+    const settings = Config.initSettings(configJSON.settings, flags);
 
-    // Store the directory of the config path, if one was provided.
-    this._configDir = configPath ? path.dirname(configPath) : undefined;
+    // Augment passes with necessary defaults and require gatherers.
+    const passesWithDefaults = Config.augmentPassesWithDefaults(configJSON.passes);
+    Config.adjustDefaultPassForThrottling(settings, passesWithDefaults);
+    const passes = Config.requireGatherers(passesWithDefaults, configDir);
 
     /** @type {LH.Config['settings']} */
-    this._settings = configJSON.settings || {};
+    this.settings = settings;
     /** @type {LH.Config['passes']} */
```

Comment:
i think these casts arent necessary any longer?

Reply:
Well, they're kind of casts, but they're not coercive, so if the lhs can't be treated as these things there will be an error. That's a nice thing to have because if future changes to `config.js` accidentally change the lhs type, the error will be here. If not, the class's property type will just be whatever the accidental type is, and the error will instead be deep somewhere else where the property is actually used. It's also nice for file-level documentation since these are pretty much the Config instance API.

Some of this (and some of `config.d.ts`) can be deleted when typescript supports `@implements`, though (https://github.com/Microsoft/TypeScript/issues/17498)